File: selenium-agent-mcp-starter-python/src/agents/test_generator_agent.py
import json
from pathlib import Path

from selenium.webdriver.common.by import By

from src.agents.base_agent import BaseAgent
from src.config.driver_factory import DriverFactory
from src.pages.login_page import LoginPage


class TestGeneratorAgent(BaseAgent):
    def execute(self, task_definition_path: str):
        repo_root = Path(__file__).resolve().parents[2]
        map_path = (repo_root / task_definition_path).resolve()

        with open(map_path, "r", encoding="utf-8") as file:
            raw_content = file.read()
            agent_map = json.loads(self._interpolate_env(raw_content))

        self._generate_artifacts(repo_root, agent_map["tasks"])
        self._execute_tasks(agent_map["tasks"])

    def _execute_tasks(self, tasks):
        driver = DriverFactory.create_driver()

        try:
            for task in tasks:
                driver.get(task["url"])

                for action in task["actions"]:
                    action_type = action["type"]

                    if action_type == "navigate":
                        driver.get(action.get("url", task["url"]))
                    elif action_type == "login":
                        login_page = LoginPage(driver)
                        login_page.login(action.get("username", ""), action.get("password", ""))
                    elif action_type == "click":
                        driver.find_element(By.CSS_SELECTOR, action.get("selector", "body")).click()
                    elif action_type == "type":
                        element = driver.find_element(By.CSS_SELECTOR, action.get("selector", "body"))
                        element.clear()
                        element.send_keys(action.get("value", ""))
                    elif action_type == "assertText":
                        body_text = driver.find_element(By.TAG_NAME, "body").text
                        expected = action.get("value") or action.get("text", "")
                        if expected not in body_text:
                            raise AssertionError(f"Expected page to contain '{expected}'")
                    elif action_type == "assertUrl":
                        current_url = driver.current_url
                        expected_url = action.get("url", "")
                        if expected_url not in current_url:
                            raise AssertionError(f"Expected URL '{current_url}' to include '{expected_url}'")
                    elif action_type == "screenshot":
                        screenshot_path = Path(action.get("path", "./screenshots/generated.png"))
                        screenshot_path.parent.mkdir(parents=True, exist_ok=True)
                        driver.save_screenshot(str(screenshot_path))
        finally:
            DriverFactory.quit_driver(driver)

    def _generate_artifacts(self, repo_root: Path, tasks):
        output_dir = repo_root / "tests" / "generated"
        report_path = repo_root / "docs" / "generated-coverage.md"
        output_dir.mkdir(parents=True, exist_ok=True)
        report_path.parent.mkdir(parents=True, exist_ok=True)

        (output_dir / "test_ai_generated.py").write_text(self._render_pytest(tasks), encoding="utf-8")
        report_path.write_text(self._render_coverage(repo_root, tasks), encoding="utf-8")

    def _render_pytest(self, tasks) -> str:
        test_blocks = []
        for index, task in enumerate(tasks):
            name = self._to_snake_case(task.get("name", f"generated_task_{index + 1}"))
            lines = [
                f"def test_{name}(driver):",
                f"    login_page = LoginPage(driver)",
                f"    login_page.navigate_to({task['url']!r})",
            ]
            lines.extend(self._render_step(action) for action in task["actions"])
            test_blocks.append("\n".join(lines))

        return "\n\n".join(
            [
                "from selenium.webdriver.common.by import By",
                "from src.pages.login_page import LoginPage",
                "",
                *test_blocks,
                "",
            ]
        )

    def _render_step(self, action) -> str:
        action_type = action["type"]
        if action_type == "navigate":
            return f"    login_page.navigate_to({action.get('url', '/')!r})"
        if action_type == "login":
            return f"    login_page.login({action.get('username', '')!r}, {action.get('password', '')!r})"
        if action_type == "click":
            return f"    login_page.click((By.CSS_SELECTOR, {action.get('selector', 'body')!r}))"
        if action_type == "type":
            return f"    login_page.type((By.CSS_SELECTOR, {action.get('selector', 'body')!r}), {action.get('value', '')!r})"
        if action_type == "assertText":
            expected = action.get("value") or action.get("text", "")
            return f"    assert {expected!r} in driver.find_element(By.TAG_NAME, 'body').text"
        if action_type == "assertUrl":
            return f"    assert {action.get('url', '')!r} in driver.current_url"
        if action_type == "screenshot":
            return f"    driver.save_screenshot({action.get('path', './screenshots/generated.png')!r})"
        return "    # Unsupported action"

    def _render_coverage(self, repo_root: Path, tasks) -> str:
        existing_tests = sorted(
            str(path.relative_to(repo_root)).replace("\\", "/") for path in (repo_root / "tests").rglob("test_*.py")
        )
        missing = []
        for task in tasks:
            task_name = self._to_snake_case(task.get("name", task["url"]))
            if not any(task_name in existing_test for existing_test in existing_tests):
                missing.append(task.get("name", task["url"]))

        lines = ["## Selenium Python Coverage Report", "", "### Covered"]
        lines.extend(f"- [x] {test_file}" for test_file in existing_tests)
        lines.extend(["", "### Missing"])
        if missing:
            lines.extend(f"- [ ] {item}" for item in missing)
        else:
            lines.append("- [x] No obvious gaps detected")
        lines.append("")
        return "\n".join(lines)

    def _to_snake_case(self, value: str) -> str:
        return "_".join(part for part in "".join(ch.lower() if ch.isalnum() else " " for ch in value).split())

    def _interpolate_env(self, content: str) -> str:
        import os
        import re

        return re.sub(r"\$\{([A-Z0-9_]+)\}", lambda match: os.getenv(match.group(1), ""), content)

File: selenium-agent-mcp-starter-java/src/main/java/com/example/agents/TestGeneratorAgent.java
package com.example.agents;

import com.example.config.DriverFactory;
import com.example.pages.LoginPage;
import com.example.utils.JsonUtils;
import com.fasterxml.jackson.databind.JsonNode;
import org.openqa.selenium.By;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.TakesScreenshot;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;

import java.io.File;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class TestGeneratorAgent implements Agent {
    private WebDriver driver;

    @Override
    public void execute(String mapFilePath) throws Exception {
        Path workingDir = Paths.get("").toAbsolutePath();
        Path repoRoot = workingDir.endsWith("selenium-agent-mcp-starter-java")
            ? workingDir
            : workingDir.resolve("selenium-agent-mcp-starter-java");
        Path absoluteMapPath = repoRoot.resolve(mapFilePath).normalize();
        JsonNode root = JsonUtils.parseJsonFile(absoluteMapPath.toString());
        JsonNode tasks = root.get("tasks");

        generateArtifacts(repoRoot, tasks);
        driver = DriverFactory.createDriver();

        try {
            for (JsonNode task : tasks) {
                String url = task.get("url").asText();
                driver.get(url);

                for (JsonNode action : task.get("actions")) {
                    String type = action.get("type").asText();

                    switch (type) {
                        case "navigate":
                            driver.get(action.path("url").asText(url));
                            break;
                        case "login":
                            LoginPage loginPage = new LoginPage(driver);
                            loginPage.login(action.path("username").asText(""), action.path("password").asText(""));
                            break;
                        case "click":
                            driver.findElement(By.cssSelector(action.path("selector").asText("body"))).click();
                            break;
                        case "type":
                            WebElement input = driver.findElement(By.cssSelector(action.path("selector").asText("body")));
                            input.clear();
                            input.sendKeys(action.path("value").asText(""));
                            break;
                        case "assertText":
                            String bodyText = driver.findElement(By.tagName("body")).getText();
                            String expectedText = action.has("value") ? action.get("value").asText() : action.path("text").asText("");
                            if (!bodyText.contains(expectedText)) {
                                throw new AssertionError("Expected page to contain: " + expectedText);
                            }
                            break;
                        case "assertUrl":
                            String expectedUrl = action.path("url").asText("");
                            String currentUrl = driver.getCurrentUrl();
                            if (!currentUrl.contains(expectedUrl)) {
                                throw new AssertionError("Expected URL '" + currentUrl + "' to include '" + expectedUrl + "'");
                            }
                            break;
                        case "screenshot":
                            Path screenshotPath = repoRoot.resolve(action.path("path").asText("screenshots/generated.png")).normalize();
                            Files.createDirectories(screenshotPath.getParent());
                            File screenshotFile = ((TakesScreenshot) driver).getScreenshotAs(OutputType.FILE);
                            Files.copy(screenshotFile.toPath(), screenshotPath, StandardCopyOption.REPLACE_EXISTING);
                            break;
                        default:
                            break;
                    }
                }
            }
        } finally {
            DriverFactory.quitDriver();
        }
    }

    private void generateArtifacts(Path repoRoot, JsonNode tasks) throws Exception {
        Path outputDir = repoRoot.resolve("src/test/java/com/example/tests/generated");
        Path reportPath = repoRoot.resolve("docs/generated-coverage.md");
        Files.createDirectories(outputDir);
        Files.createDirectories(reportPath.getParent());

        Files.writeString(outputDir.resolve("GeneratedJourneysTest.java"), renderJUnit(tasks));
        Files.writeString(reportPath, renderCoverage(repoRoot, tasks));
    }

    private String renderJUnit(JsonNode tasks) {
        List<String> tests = new ArrayList<>();

        for (int i = 0; i < tasks.size(); i++) {
            JsonNode task = tasks.get(i);
            String testName = toMethodName(task.path("name").asText("generated task " + (i + 1)));
            List<String> lines = new ArrayList<>();
            lines.add("    @Test");
            lines.add("    void " + testName + "() {");
            lines.add("        loginPage.navigateTo(\"" + task.get("url").asText() + "\");");

            for (JsonNode action : task.get("actions")) {
                lines.add(renderStep(action));
            }

            lines.add("    }");
            tests.add(String.join("\n", lines));
        }

        return String.join("\n", List.of(
            "package com.example.tests.generated;",
            "",
            "import com.example.config.DriverFactory;",
            "import com.example.pages.LoginPage;",
            "import org.junit.jupiter.api.*;",
            "import org.openqa.selenium.By;",
            "import org.openqa.selenium.WebDriver;",
            "",
            "import static org.junit.jupiter.api.Assertions.assertTrue;",
            "",
            "class GeneratedJourneysTest {",
            "    private WebDriver driver;",
            "    private LoginPage loginPage;",
            "",
            "    @BeforeEach",
            "    void setUp() {",
            "        driver = DriverFactory.createDriver();",
            "        loginPage = new LoginPage(driver);",
            "    }",
            "",
            "    @AfterEach",
            "    void tearDown() {",
            "        DriverFactory.quitDriver();",
            "    }",
            "",
            String.join("\n\n", tests),
            "}",
            ""
        ));
    }

    private String renderStep(JsonNode action) {
        String type = action.get("type").asText();
        switch (type) {
            case "navigate":
                return "        loginPage.navigateTo(\"" + action.path("url").asText("/") + "\");";
            case "login":
                return "        loginPage.login(\"" + action.path("username").asText("") + "\", \"" + action.path("password").asText("") + "\");";
            case "click":
                return "        driver.findElement(By.cssSelector(\"" + action.path("selector").asText("body") + "\")).click();";
            case "type":
                return String.join("\n", List.of(
                    "        {",
                    "            var input = driver.findElement(By.cssSelector(\"" + action.path("selector").asText("body") + "\"));",
                    "            input.clear();",
                    "            input.sendKeys(\"" + action.path("value").asText("") + "\");",
                    "        }"
                ));
            case "assertText":
                String expectedText = action.has("value") ? action.get("value").asText() : action.path("text").asText("");
                return "        assertTrue(driver.findElement(By.tagName(\"body\")).getText().contains(\"" + expectedText + "\"));";
            case "assertUrl":
                return "        assertTrue(driver.getCurrentUrl().contains(\"" + action.path("url").asText("") + "\"));";
            case "screenshot":
                return "        // Screenshot path: " + action.path("path").asText("screenshots/generated.png");
            default:
                return "        // Unsupported action";
        }
    }

    private String renderCoverage(Path repoRoot, JsonNode tasks) throws Exception {
        List<String> existingTests = Files.exists(repoRoot.resolve("src/test/java"))
            ? Files.walk(repoRoot.resolve("src/test/java"))
                .filter(path -> path.toString().endsWith("Test.java"))
                .map(path -> repoRoot.relativize(path).toString().replace("\\", "/"))
                .sorted()
                .collect(Collectors.toList())
            : List.of();

        List<String> missing = new ArrayList<>();
        for (JsonNode task : tasks) {
            String taskName = toMethodName(task.path("name").asText(task.get("url").asText()));
            boolean present = existingTests.stream().anyMatch(test -> test.toLowerCase().contains(taskName.toLowerCase()));
            if (!present) {
                missing.add(task.path("name").asText(task.get("url").asText()));
            }
        }

        List<String> lines = new ArrayList<>();
        lines.add("## Selenium Java Coverage Report");
        lines.add("");
        lines.add("### Covered");
        for (String test : existingTests) {
            lines.add("- [x] " + test);
        }
        lines.add("");
        lines.add("### Missing");
        if (missing.isEmpty()) {
            lines.add("- [x] No obvious gaps detected");
        } else {
            for (String item : missing) {
                lines.add("- [ ] " + item);
            }
        }
        lines.add("");
        return String.join("\n", lines);
    }

    private String toMethodName(String value) {
        String cleaned = value.replaceAll("[^A-Za-z0-9]+", " ").trim();
        String[] parts = cleaned.split("\\s+");
        StringBuilder builder = new StringBuilder("test");
        for (String part : parts) {
            if (part.isEmpty()) {
                continue;
            }
            builder.append(part.substring(0, 1).toUpperCase()).append(part.substring(1));
        }
        return builder.toString();
    }
}
